src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs

```diff
@@ -1085,291 +1047,200 @@ internal static string GetExpandedEnvironmentVariable(string name, EnvironmentVa
+            {
+                // %PSModulePath% is the same so return the cache.
+                return s_modulePathCache;
+            }
 
-            if (!string.IsNullOrEmpty(newModulePathString))
+            lock (s_syncPSModulePath)
             {
-#if !UNIX
-                // If on Windows, we want to add the System32 Windows PowerShell module directory
-                // so that Windows modules are discoverable
-                string windowsPowerShellModulePath = GetWindowsPowerShellPSHomeModulePath();
-                if (!newModulePathString.Contains(windowsPowerShellModulePath, StringComparison.OrdinalIgnoreCase))
+                if (string.Equals(s_psModulePathEnvVar, currentModulePath, StringComparison.Ordinal) &&
+                    string.IsNullOrEmpty(InternalTestHooks.TestWindowsPowerShellPSHomeLocation) &&
+                    InternalTestHooks.TestPowerShellPSModulePaths?.Length == 0)
                 {
-                    newModulePathString += Path.PathSeparator + windowsPowerShellModulePath;
+                    // Another thread already processed %PSModulePath%.
+                    return s_modulePathCache;
                 }
+
+                // %PSModulePath% is changed so parse it in the thread and update module path cache.
+                s_modulePath = new List<string>();
+
+                string allUsersModulePath = PowerShellConfig.Instance.GetModulePath(ConfigScope.AllUsers);
+                string personalModulePath = PowerShellConfig.Instance.GetModulePath(ConfigScope.CurrentUser);
+
+                AddStdModulePath(currentModulePath, allUsersModulePath, personalModulePath);
+
+                if (s_modulePath.Count != 0)
+                {
+#if !UNIX
+                    // If on Windows, we want to add the System32 Windows PowerShell module directory
+                    // so that Windows modules are discoverable
+                    string windowsPowerShellModulePath = GetWindowsPowerShellPSHomeModulePath();
+                    var insertIndex = IndexOfPath(windowsPowerShellModulePath);
+                    if (insertIndex == -1)
+                    {
+                        s_modulePath.Add(windowsPowerShellModulePath);
+                    }
 #endif
+                }
 
-                // Set the environment variable...
-                Environment.SetEnvironmentVariable(Constants.PSModulePathEnvVar, newModulePathString);
-            }
+                s_modulePathCache = s_modulePath;
+                s_psModulePathEnvVar = currentModulePath;
 
-            return newModulePathString;
+                if (InternalTestHooks.TestPowerShellPSModulePaths?.Length > 0)
+                {
+                    var expectedPaths = InternalTestHooks.TestPowerShellPSModulePaths;
+                    for (int i = 0; i < expectedPaths.Length; i++)
+                    {
+                        var expectedPath = expectedPaths[i].ToString();
```

Comment:
```suggestion
                        string expectedPath = expectedPaths[i].ToString();
```

Reply:
I believe ToString() implies the return type.